Trainer refactoring

justusschock opened the pull request:
This is a first draft to refactor trainer (combine code where possible) and introduce a predictor.

The metric logging was moved from the networks closure to the trainer.

@ORippler : maybe we could also merge experiments and rename the `AbstractTrainer` to `BaseTrainer` since it is not abstract anymore?

Also some tf tests fail due to shape missmatch. Have you any idea why?

@mibaumgartner : does this match your idea of a predictor? Do you have any improvements compared to your own prototype?


fix #39 
fix #46 ?

EDIT: Docstrings are still missing

justusschock committed:
Start refactoring trainer and experiment

justusschock committed:
make new torch trainer and experiment working

mibaumgartner commented on delira/training/predictor.py:
```diff
@@ -0,0 +1,208 @@
+            return new_val_score < old_val_score
+
+    @staticmethod
+    def calc_metrics(groundtruths, *predictions, metrics={}, metric_keys=None):
```
`calc_metrics` is called in the network trainer without the option to change the metrics_keys. 

https://github.com/justusschock/delira/blob/722cd55dd308f2ff067f34d5eb323511fe3073c2/delira/training/abstract_trainer.py#L334-L338

It may be useful to introduce a new argument in the trainer to set the metrics_keys so it is not always necessary to overwrite the function if different metric keys are needed.

mibaumgartner commented on delira/training/predictor.py:
```diff
@@ -0,0 +1,208 @@
+
+        batchgen._finish()
+
+        predictions_all = zip(*predictions_all)
```
For `zip` to work properly it is necessary that `predictions_all` is a list of lists. For some networks (like detection networks) which complete multiple tasks (detection, segmentation, classification) I like to use a dict for my network output (so I don't have to remember which entry belongs to which task). Is there any possibility to allow that here?

justusschock replied:
We would have to entirely switch to a dict, because maintaining both possibilities would be an unnecessary overhead I guess.

What do you think @haarburger @ORippler ?

ORippler replied:
I think we shouldn't mix `lists` and `dicts`. In current master, we only use lists for `labels_all` and `outputs_all`, but now we intermingle it already now with `predictions_all` and `total_input_dict`. If we need the whole information in `total_input_dict` in order to calculate proper validation metrics, we should switch to dicts totally.

justusschock replied:
I think, I'd like to move this to another issue, since this issue shouldn't change anything on the user-side, but switching from lists to dicts will.

mibaumgartner commented on delira/training/abstract_trainer.py:
```diff
@@ -1,31 +1,57 @@
+                 losses: dict,
+                 optimizer_cls: type,
+                 optimizer_params: dict,
+                 train_metrics: dict,
```
Maybe allow `train_dataset_metrics`?

justusschock replied:
This was not allowed on purpose, because this would need the whole trainset to be cached, which is often not a good idea.

mibaumgartner replied:
For classification and detection it is only needed to cache the label and bounding boxes, which is pretty reasonable. Nevertheless, I get why it's not in there but I think it would be nice to have the option.

justusschock replied:
@ORippler do you think, we should have this option for the trainset as well?

ORippler replied:
What benefit would `train_dataset_metrics` bring us?

mibaumgartner commented on delira/training/metrics.py:
```diff
@@ -1,106 +1,37 @@
-        def __init__(self):
-            super().__init__()
+        return self._score_fn(**arrays, **kwargs)
 
```
The AUC score would be a nice addtional feature. If needed, I can provide an implementation later today or tomorrow.

justusschock replied:
Feel free to do so.

ORippler replied:
@mibaumgartner please resolve if done or no longer planned

mibaumgartner replied:
Sorry, I didn't get any notification. I already sketched a first version, gonna test it during the next days (maybe even today) and then create a pull request.

justusschock committed:
Merge branch 'master' into trainer_refactoring

justusschock committed:
Update pytorch_trainer.py

justusschock committed:
Update pytorch_trainer.py

ORippler commented on delira/training/predictor.py:
```diff
@@ -0,0 +1,208 @@
+            super().__setattr__(key, value)
+
+    @staticmethod
+    def _is_better_val_scores(old_val_score, new_val_score,
```
from what this function does, it should live in the `AbstractTrainer` and not in the predictor imo. It is so far also only used by the `AbstractTrainer`

ORippler commented on delira/training/predictor.py:
```diff
@@ -0,0 +1,208 @@
+            )
+        )
+
+    def predict_data_mgr(self, datamgr, batchsize=None, verbose=False):
```
in tf, a singleton dimension is added somewhere in here, leading to shape missmatches for inputs

ORippler replied:
Singleton is also added for pytorch presumably, as `classification_pytorch.ipynb` fails with the following error message:

```RuntimeError: Expected 4-dimensional input for 4-dimensional weight [64, 1, 7, 7], but got 5-dimensional input of size [64, 1, 1, 224, 224] instead```

We again have a singleton dimension (and some additional shape missmatch) when calling `predictor.predict_data_mgr`
Real question is though: why are our tests not catching this?

ORippler reviewed (changes requested):
will investigate shape errors tomorrow. Furthermore, experiment.stratified_kfold_predict has to be adapted (might also be related to the shape errors)

ORippler commented on delira/training/predictor.py:
```diff
@@ -0,0 +1,208 @@
+                            batch_dict[key] = [val]
+
+                for key, val_list in batch_dict.items():
+                    batch_dict[key] = np.array(val_list)
```
why this got changed from [np.concatenate](https://github.com/justusschock/delira/blob/b663d982d87f3a0be7b346d25576770b184d4db9/delira/training/tf_trainer.py#L463) to np.array? should fix the singleton issue presumably:

justusschock replied:
My mistake. Will fix this probably tomorrow

mibaumgartner committed:
Merge branch 'trainer_refactoring' of https://github.com/justusschock/delira into trainer_refactoring

mibaumgartner committed:
First version of auc metric implementation

justusschock committed:
Resolve DImension Error

justusschock committed:
Move `_is_better_val_score` from predictor to abstract trainer

justusschock committed:
Fix error in torchvision datasets due to latest torchvision release

justusschock committed:
Fix error due to missing brackets for file extension checks

justusschock committed:
Add docstrings

justusschock committed:
Update pytorch_trainer docstrings

mibaumgartner committed:
Merge remote-tracking branch 'upstream/trainer_refactoring' into trainer_refactoring

mibaumgartner committed:
Finished auc metric and added doc strings

mibaumgartner committed:
Added unit test for metric wrapper and auc metric

mibaumgartner committed:
Remove dummy experiment file

mibaumgartner committed:
Remove todo

mibaumgartner committed:
Added docstring

mibaumgartner committed:
Updated docstring

justusschock committed:
Update test_metrics.py

justusschock committed:
Merge branch 'master' into trainer_refactoring

justusschock committed:
Merge pull request #73 from mibaumgartner/trainer_refactoring

AUC metric

ORippler committed:
resolve merge conflicts and merge

ORippler committed:
update TfExperiment.test, change test signature

ORippler reviewed (changes requested):
Okay tests seem to be passing so far. Further ToDOs/Open Questions:

* [x] Write/Update Docstrings for all classes
* [x] Write Unittests for `Predictor` class
* [x] Discuss/decide on whether conversion from list of list to list of dicts should be part of this PR or another one
* [x] Argue about the usability of training dataset metrics

justusschock committed:
Make consistent logging

justusschock committed:
Continue merging trainers

justusschock committed:
update tests

ORippler committed:
fix feed errors by using default placeholder

justusschock committed:
Change list outputs to dict outputs

justusschock committed:
Change networks to return dicts

justusschock committed:
remove duplicate argument in tf trainer

justusschock committed:
deprecate models in favor of to be announced repo

justusschock committed:
Merge branch 'master' into trainer_refactoring

justusschock committed:
start updating notebooks

justusschock committed:
Start Experiment refactoring

justusschock committed:
Update Experiment Tests and Fix Bugs

justusschock committed:
Fix Naming issue

Fix naming issue in kfold

justusschock committed:
continue test fixing

justusschock committed:
Merge branch 'trainer_refactoring' of https://github.com/justusschock/delira into trainer_refactoring

justusschock committed:
add docstrings

justusschock committed:
fix minor bugs, change descriptor during validation phase and introduce validation frequency

mibaumgartner commented on delira/training/experiment.py:
```diff
@@ -1,388 +1,582 @@
-    def __init__(self, n_epochs, *args, **kwargs):
+
+    def __init__(self,
+                 params: Parameters,
```
Is it possible to make the `params` object optional in the `__init__`? (It would change the signature and thus backwards compatibility could be problematic)

justusschock replied:
We would have to change the argument order here and I'm not sure if I'd like that, because IMO, the parameters are more important than other stuff and I can't really imagine a scenario where you wouldn't pass parameters here. What do you think @ORippler?

ORippler replied:
I also don't know of a situation where you wouldn't pass params to `__init__`. Though I think we should rethink/define on what to pass as additional kwarg and what to include inside the params.

Idea: Everything that can be tuned/varied during training/testing should go into params, as they are intended to be utilized for the hyperparameter search)

mibaumgartner replied:
I encountered multiple situations where I wanted to initialise the experiment to get the `savepath` without initialising the parameters, but maybe that is only me ^^

ORippler replied:
with current implementation, it must be typing.Union[str, Parameters]

justusschock replied:
will fix this

mibaumgartner commented on delira/training/experiment.py:
```diff
@@ -1,388 +1,582 @@
+        key_mapping : dict
+            mapping between data_dict and model inputs (necessary for
+            prediction with :class:`Predictor`-API)
+        val_score_key : str or None
```
I think  it's more intuitive if `val_score_key` (currently set in `__init__`) and `val_score_mode`(currently set in `run`) would be set in the same function (maybe in `run`). What do you think about this?

justusschock replied:
I think I left it that way, because I think this is what it was like until now, but I agree

mibaumgartner commented on delira/training/tf_trainer.py:
```diff
@@ -60,55 +85,52 @@ def __init__(self, network, save_path,
         for key, val in kwargs.items():
             setattr(self, key, val)
 
     def _setup(self, network, optim_fn, optimizer_cls, optimizer_params,
```
It seems like the `PytorchNetworkTrainer` already supports the `resume` function of the `experiment` class and loads the latest checkpoint. Currently, it is a bit misleading if only one trainer supports this functionality. In general, I'm a big fan to have the option to resume the training.

ORippler replied:
we should imo include a switch on wether to restore from best model or latest model.
Also, for `TfNetworkTrainer`, `start_epoch` still has to be updated accordingly

justusschock replied:
I think the resume inside the trainer should only restore the latest model, since we want t continue training here...

ORippler replied:
okay, still we should update `start_epoch` then

mibaumgartner committed:
Fix some bugs, torch tests should complete now

justusschock committed:
Add resume to tf trainer

justusschock committed:
merge master into trainer_refactoring

justusschock committed:
fix merge

justusschock committed:
fix indent tf trainer

justusschock committed:
switch to suitable permutation

ORippler committed:
specify `data` in signature as dict type

ORippler committed:
switch from list to dicts for tf

ORippler committed:
pep8

ORippler commented on delira/training/experiment.py:
```diff
@@ -1,388 +1,583 @@
-    def setup(self, *args, **kwargs):
-        """
-        Abstract Method to setup a :class:`AbstractNetworkTrainer`
+        # params could also be a file containing a pickled instance of parameters
```
if we accept filepaths for params, this must come before handling `n_epochs is None`

ORippler commented on delira/training/predictor.py:
```diff
@@ -0,0 +1,277 @@
+    __KEYS_TO_GUARD = []
+
+    def __init__(self, model, key_mapping: dict, 
+                 convert_batch_to_npy_fn=lambda x: x,
```
lambda x: x no longer captures the desired default behaviour (which seems to be [discard *args and pass **kwargs) of returned model.call()](https://github.com/justusschock/delira/blob/trainer_refactoring/delira/training/predictor.py#L113). Either make it positional arg or change default setting. Also, the naming is no longer fitting (we convert the tensors inside the dict entries to np.ndarrays if required).

ORippler commented on delira/models/classification/classification_network_tf.py:
```diff
@@ -181,11 +188,7 @@ def closure(model: typing.Type[AbstractTfNetwork], data_dict: dict,
         #logging.info({'image_grid': {"image_array": inputs, "name": image_names,
         #                             "title": "input_images", "env_appendix": "_%02d" % fold}})
-        return metric_vals, loss_vals, [preds]
+        return metric_vals, loss_vals, {"pred": preds}
```
is this still required? As far as i can see, the dict structure is only required to pass it through Predictor._convert_batch_to_npy(). My outputs are now naturally of dict type and therefore I would pass the whole returned dict through here.

justusschock replied:
No, this isn't required, This was only my first try to make it work with TF :D

ORippler commented on delira/training/predictor.py:
```diff
@@ -0,0 +1,277 @@
+        if metric_keys is None:
+            metric_keys = {k: ("pred", "label") for k in metrics.keys()}
+
+        return {key: metric_fn(batch_dict[metric_keys[key][0]], 
```
we should switch to allow variable length for metric_keys

ORippler commented on delira/training/predictor.py:
```diff
@@ -0,0 +1,277 @@
+        )[1]
+
+    def predict_data_mgr(self, datamgr, batchsize=None, metrics={},
+                         metric_keys=None, verbose=False):
```
Should we add the ability to compute metrics over the whole data_mgr? Instead of returning a list of batch_level metrics in addiction to all predictions. I think we had decided not to for training data_managers...
Also, description of what is returned is missing

justusschock replied:
We had decided to do this in general but to postpone this, until we have a way to do this lazy and non-lazy, because for large datasets (e.g. imagenet) we might also break the lazy behavior with the validation/testset.

Will update the docstrings tomorrow together with your other comments

ORippler committed:
change to dicts

justusschock committed:
Adress requested changes: update start_epoch for tf_trainer, support arbitrary number of metric arguments in predictor and fix type annotation of params in experiment

justusschock committed:
fix indent causing swigpy pickling error

justusschock committed:
fix tf experiment test and type annotations

justusschock committed:
detach network outputs, remove params from test in TF Experiment

ORippler committed:
key 'prediction' to 'pred'

ORippler committed:
prediction to pred

justusschock committed:
formulate default fn explicitly

justusschock committed:
rename argument in docstrings

justusschock committed:
Merge branch 'trainer_refactoring' of https://github.com/justusschock/delira into trainer_refactoring

justusschock committed:
revert move to predictions key

justusschock committed:
check arrays for zero-size and reshape if necessary

justusschock committed:
fix start_epoch

ORippler commented on delira/training/predictor.py:
```diff
@@ -0,0 +1,286 @@
+                for key, val_list in batch_dict.items():
+                    batch_dict[key] = np.concatenate(val_list)
+
+                preds = self.predict(batch_dict)
```
do we allow nested dicts? or not

justusschock replied:
you mean as input? This shouldn't matter, as long as your network/`prepare_batch` handle this correctly. 

For the output this becomes problematic, since we do some concatenation based on these values and we can't concatenate dicts. Another possibility would be to flatten these dicts by something like this:

```python
import collections

def flatten(d, parent_key='', sep='_'):
    items = []
    for k, v in d.items():
        new_key = parent_key + sep + k if parent_key else k
        if isinstance(v, collections.MutableMapping):
            items.extend(flatten(v, new_key, sep=sep).items())
        else:
            items.append((new_key, v))
    return dict(items)
```

This would be a design decision, we'd have to make now (as this currently causes the tests to fail). I don't really have an opinion on this. What do you think @ORippler @mibaumgartner @haarburger ?

ORippler replied:
Yeah i also identified that as the reason for failure.

I dislike the idea of flattening the dict, as the new keys would need to be passed to `calc_metrics` and the user informed of this.

I would prefer a recursion-based solution to allow for nested dicts, though `calc_metrics` would also need to be updated (currently, the values used for metric calculation need to lie on the top-level of the dict).

mibaumgartner replied:
I'm not sure if it is beneficial to support nested dicts. Are there any useful scenarios where this could/must be used?

justusschock replied:
I wouldn't like a recursion based approach. I can't see any benefits from it, but if we need to resolve arbitrary recursions multiple times this is bad style IMHO

ORippler replied:
@mibaumgartner the reason we run into that failure is because [tf.Session.run()](https://www.tensorflow.org/api_docs/python/tf/Session#run) allows nested dicts and i used that functionality already inside `ClassificationNetworkBase` (it makes the logging of training loss easier as we have model.closure as static function for pytorch's multi-gpu approach). Could handle the unpacking inside the classes' `__call__` though.

justusschock replied:
@ORippler I'm pro nested dict, if we can do this by resolving the recursion only once (i.e. something similar like flattening). What I don't like: If we would have to resolve the recursion mutiple times (i.e. one time for concatenation and another time for metric calculation (and maybe a third time for logging/reporting back to HP search in the future). This gets extremly inefficient quite fast.

Another problem would be, that we could have the same key in multiple subdicts and it is unclear which item the key refers to (e.g. in metric calculation).

justusschock replied:
@mibaumgartner @ORippler I would like to move this on. What are we gonna do here?

mibaumgartner replied:
If it is benficial for tf applications, we should implement a soultion....

justusschock replied:
But which one? The only one, I can think of, resolving the possible ambiguity in keys would be the one above. Alternatively we could wrap in in a `LookupConfig` (the parameters baseclass) and use `nested_get` here as well

justusschock replied:
So, I just implemented the version with `LookupConfig` and nested dicts.

justusschock committed:
added additional checks

justusschock committed:
Make dict a lookup config to support nested dicts via `nested_get`

justusschock committed:
Make item concat working for nested dicts

justusschock committed:
Pin scipy requirement

Trixi currently has a problem with their dependencies and imports not matching and thus breaking. Should be fixed for now by pinning scipy (we don't use the other problematic dependencies)

justusschock committed:
Remove unnecessary imports

justusschock committed:
Merge branch 'trainer_refactoring' of https://github.com/justusschock/delira into trainer_refactoring

justusschock committed:
Merge branch 'master' into trainer_refactoring

justusschock committed:
Update docs

mibaumgartner committed:
Add kwargs in predictor for prepare_batch_fn

justusschock committed:
Convert batchdict back to numpy for metric calculation and introduce lazy  behavior

justusschock committed:
Merged origin commits into offline branch

justusschock committed:
fix generator behavior

justusschock committed:
make experiment.test return the first generator item

justusschock committed:
Merge branch 'master' into trainer_refactoring

justusschock committed:
add kwargs to overwritten predict_data_mgr functions

justusschock committed:
merge Parallel_master

Commit:
PEP-8 Auto-Fix

justusschock committed:
Merge GAN

justusschock committed:
Merge PEP-8 Autofix

justusschock committed:
Add style fixes and common function to search for previous checkpoints

justusschock committed:
fix infinite recursion by hard type checking instead of instance checking

justusschock committed:
fix pep8

justusschock committed:
Remove TrixiExperiment as Experiment Baseclass

justusschock committed:
correct indent in tf trainer

justusschock committed:
shallow copy

Make shallow copy of batchdict to retain keys, which might get popped in `prepare_batch`

justusschock committed:
initialize uninitialized members in TfExeriment.Test

justusschock committed:
remove param argument from test for TfExperiment.test

mibaumgartner reviewed (approved):
Implementation looks good :+1: 
We should include the `Predictor` inside the `training.rst` so it is displayed correctly in the documentation.
I think that should be a quick fix, so i approve the changes.

justusschock committed:
add predictor to docs
